File: instabiz/instabiz/report/ib_bank_payment_file/ib_bank_payment_file.py
"""IB Bank Payment File — salaries or vendor payments as one bank bulk-upload file.
The table shows every payee and why a line is left out (no account / IFSC);
the Download button writes the HDFC upload CSV (instabiz.overrides.bank_payment_file)."""
import frappe
from frappe import _

from instabiz.overrides.bank_payment_file import ROLES, rows_for


def execute(filters=None):
	f = frappe._dict(filters or {})
	frappe.only_for(ROLES)
	rows = rows_for(f.source or "Salary", f.month, f.from_date, f.to_date)
	cols = [
		{"label": _("Voucher"), "fieldname": "voucher", "fieldtype": "Dynamic Link", "options": "source", "width": 170},
		{"label": _("Source"), "fieldname": "source", "fieldtype": "Data", "hidden": 1},
		{"label": _("Payee"), "fieldname": "name", "fieldtype": "Data", "width": 200},
		{"label": _("Amount"), "fieldname": "amount", "fieldtype": "Currency", "width": 120},
		{"label": _("Account No"), "fieldname": "account", "fieldtype": "Data", "width": 150},
		{"label": _("IFSC"), "fieldname": "ifsc", "fieldtype": "Data", "width": 110},
		{"label": _("Mode"), "fieldname": "mode", "fieldtype": "Data", "width": 60},
		{"label": _("Left out because"), "fieldname": "problem", "fieldtype": "Data", "width": 200},
	]
	ok = [r for r in rows if not r.problem]
	summary = [
		{"label": _("In file"), "value": len(ok), "datatype": "Int", "indicator": "green"},
		{"label": _("Amount in file"), "value": sum(r.amount for r in ok), "datatype": "Currency"},
		{"label": _("Left out"), "value": len(rows) - len(ok), "datatype": "Int", "indicator": "red" if len(rows) > len(ok) else "gray"},
	]
	return cols, rows, None, None, summary

File: instabiz/instabiz/report/ib_bank_payment_file/ib_bank_payment_file.js
frappe.query_reports["IB Bank Payment File"] = {
	filters: [
		{ fieldname: "source", label: __("Pay"), fieldtype: "Select", options: "Salary\nVendor", default: "Salary", reqd: 1 },
		{ fieldname: "month", label: __("Salary Month"), fieldtype: "Date", default: frappe.datetime.month_start(),
			depends_on: "eval:doc.source=='Salary'" },
		{ fieldname: "from_date", label: __("From Date"), fieldtype: "Date", default: frappe.datetime.get_today(),
			depends_on: "eval:doc.source=='Vendor'" },
		{ fieldname: "to_date", label: __("To Date"), fieldtype: "Date", default: frappe.datetime.get_today(),
			depends_on: "eval:doc.source=='Vendor'" },
	],
	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "problem" && data && data.problem) value = `<span class="text-danger">${value}</span>`;
		return value;
	},
	onload(report) {
		report.page.add_inner_button(__("Download Bank File"), () => {
			frappe.call({
				method: "instabiz.overrides.bank_payment_file.download",
				args: report.get_values(),
				freeze: true,
			}).then((r) => {
				const m = r.message;
				const blob = new Blob([m.content], { type: "text/csv" });
				const a = document.createElement("a");
				a.href = URL.createObjectURL(blob);
				a.download = m.filename;
				a.click();
				frappe.show_alert({ message: __("{0} payments, {1}", [m.count, format_currency(m.total)]), indicator: "green" });
			});
		}).addClass("btn-primary");
	},
};
